### DISCUSSION TITLE
"Missing Frame" behaviour

### DISCUSSION THREAD

**Discussion by SteffenDuenner** [ggarra13/mrv2#58] [Ideas]
Hi there,

mrViewer had the very nice and simple option to choose what to do with missing frames:
![image](https://github.com/ggarra13/mrv2/assets/33224667/89a8a421-8975-4392-9041-7bfbe64b9e80)

A very helpful option was to display missing frames by just repeating the previous frame with a large red "X" across it. This way you could assess the animation even while it was still being rendered to disk / incomplete. Showing black frames causes to much visual distraction.

**Comment by ggarra13** [comment]
Indeed.  That was a really cool thing to have.  I have added the option back in in the mrv2's main repository.  

The only difference with mrViewer is that it will repeat the previous frame only when playing forwards.  When playing backwards it will repeat the next frame instead.  I hope that's fine (if it is not, I'll have to bug Darby for a change in tlRender's core).

If you want to try it, let me know what platform you are in and I'll post a beta here for you to play with.

**Comment by ggarra13** [comment]
Never mind.  I managed to make it work just like mrViewer now.  Let me know if you want to use the beta.

**Comment by ggarra13** [comment]
Okay.  

** Removed links to pre-compiled betas which are out of date now.  **

**Comment by SteffenDuenner** [comment]
Wow! This is awesome. I didn't expect such a fast response last night and went offline (I live in another time zone ;) ).
And now I woke up and voila the feature is implemented.

I compiled mrv2 myself, but thanks for the builds, too.

And here my first feedback. I tested it with a simple EXR sequence that I manually deleted some random frames from:
The scratched frame option does what it's supposed to do, but the playback isn't "smooth" anymore whenever there's a missing frame. It somehow feels like it doesn't only replace the missing frame with a scratched previous one but it pauses playback for another 2-3 frames before resuming.
Furthermore, the red X doesn't always show up. And third, something happens with the alpha (?) of the scratched frame whenever the red X is added.

I quickly made some OBS captures that hopefully show the issues. This is a shabby 100 frame sequence playing back as it should:
https://github.com/ggarra13/mrv2/assets/33224667/e867d02d-3fbb-46fb-a77b-c9d373782444

This is the same sequence with frames 20, 40, 60 and 80 removed. Notice the missing X and the stuttery playback:
https://github.com/ggarra13/mrv2/assets/33224667/e1569df8-c010-43f8-bb64-9fb045a775c4

And this is me manually stepping through the frames around frame 40. The X is there as it should, but the alpha looks like it gets premultiplied on the scratched frame:
https://github.com/ggarra13/mrv2/assets/33224667/ee050e1b-2da3-4a2b-a101-5cd3bdf4c0c8

Thanks again for your great work :)

**Comment by ggarra13** [comment]
Can you upload the full sequence of exrs as a zip to google drive, mega.nz or some other upload service?  It is hard for me to see the issue on the movie files. In the .exr sequence I have I don't see any issues.

**Reply by SteffenDuenner** [reply]
Sure, I uploaded the 100 frame sequence to mega:

https://mega.nz/file/2jQ1VAKK#Xc_49Ch1JEkdjEQNMPIejxorGcqu-o3rNnuQIgLktVg

**Comment by ggarra13** [comment]
Okay.  I added an optimization when playing forwards with scratched frame active.  Compile  main HEAD again and see if it improves your playback issues.

The alpha is not premultiplied, but the color of the lines has some opacity in them (0.8).  Maybe that's what you are seeing?

**Reply by SteffenDuenner** [reply]
The "erosion" that looks like a premultiplication happens around the parts where the liquid meets the transparent background.
You can see it in the third screen capture movie I posted above. I switch between frame 39 and 40 several times and the image changes (not only the lines).
I will now recompile and check the other fix you added...

Are you sure you pushed the commit? Last update I currently see is
![image](https://github.com/ggarra13/mrv2/assets/33224667/3ce0810f-33a7-4449-9cc8-ae086602cdf9)

**Comment by ggarra13** [comment]
Yes, you are right.  I was missing pushing the commit.  Thanks for catching that.

**Reply by SteffenDuenner** [reply]
:) Perfect, I recompiled and playback is now smooth even with scratched frames. The red X now also shows up reliably.
The premultiplication issue is super-minor, it only affects scratched frames which are placeholders anyway.
Here's a short capture to show the "problem" in more detail:

https://github.com/ggarra13/mrv2/assets/33224667/67e0eee4-8495-484d-a0e7-85613e04a0ee

**Comment by ggarra13** [comment]
I pushed a new commit that should fix the premultiplication issue.

**Reply by SteffenDuenner** [reply]
Awesome. I recompiled and can confirm that this feature is now perfect 👍🏻
